crates/bevy_render/src/pipelined_rendering.rs

```diff
@@ -0,0 +1,154 @@
+use async_channel::{Receiver, Sender};
+
+use bevy_app::{App, AppLabel, Plugin, SubApp};
+use bevy_ecs::{
+    schedule::{MainThreadExecutor, StageLabel, SystemStage},
+    system::Resource,
+    world::{Mut, World},
+};
+use bevy_tasks::ComputeTaskPool;
+
+use crate::RenderApp;
+
+/// A Label for the sub app that runs the parts of pipelined rendering that need to run on the main thread.
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, AppLabel)]
+pub struct RenderExtractApp;
+
+/// Labels for stages in the [`RenderExtractApp`] sub app. These will run after rendering has started.
+#[derive(Debug, Hash, PartialEq, Eq, Clone, StageLabel)]
+pub enum RenderExtractStage {
+    /// When pipelined rendering is enabled this stage runs after the render schedule starts, but
+    /// before I/O processing and the main app schedule. This can be useful for something like
+    /// frame pacing.
+    BeforeIoAfterRenderStart,
+}
+
+/// Channel to send the render app from the main thread to the rendering thread
+#[derive(Resource)]
+pub struct MainToRenderAppSender(pub Sender<SubApp>);
+
+/// Channel to send the render app from the render thread to the main thread
+#[derive(Resource)]
+pub struct RenderToMainAppReceiver(pub Receiver<SubApp>);
+
+/// The [`PipelinedRenderingPlugin`] can be added to your application to enable pipelined rendering.
+/// This moves rendering into a different thread, so that the Nth frame's rendering can
+/// be run at the same time as the N + 1 frame's simulation.
+///
+/// ```text
+/// |--------------------|--------------------|--------------------|--------------------|
+/// | simulation thread  | frame 1 simulation | frame 2 simulation | frame 3 simulation |
+/// |--------------------|--------------------|--------------------|--------------------|
+/// | rendering thread   |                    | frame 1 rendering  | frame 2 rendering  |
+/// |--------------------|--------------------|--------------------|--------------------|
+/// ```
+///
+/// The plugin is dependent on the [`crate::RenderApp`] added by [`crate::RenderPlugin`] and so must
+/// be added after that plugin. If it is not added after, the plugin will do nothing.
+///
+/// A single frame of execution looks something like below    
+///
+/// ```text
+/// |-------------------------------------------------------------------|
+/// |         | BeforeIoAfterRenderStart | winit events | main schedule |
+/// | extract |---------------------------------------------------------|
+/// |         | extract commands | rendering schedule                   |
+/// |-------------------------------------------------------------------|
+/// ```
```

Comment:
I'm having trouble understanding this graphic. Is this implying the `extract` runs before the main schedule?

My mental model of how this works is something like this:
```mermaid
flowchart TB
    subgraph CPU1
    direction LR
    a1["BeforeIoAfterRenderStart"]-->a2["winit wvents"]
    a2-->a3["main schedule"]
    end
    subgraph CPU2
    direction LR
    b1["extract commands"]-->b2["rendering schedule"]
    end
    CPU1--extract-->CPU2
```

Reply:
![image](https://user-images.githubusercontent.com/2180432/213020786-68bd9af5-210e-4cd1-a0bd-082b64cfc351.png)